File: Java/038_CountAndSay.java
/*************************************************************************
    > File Name: 038_CountAndSay.java
    > Author: daiyu
	> Url: https://leetcode.com/problems/count-and-say/
    > Created Time: 二  9/22 14:03:43 2015
 ************************************************************************/
public class Solution {
    public String countAndSay(int n) {
        String str = "1";
        int i = 1;
        
        while (i < n) {
            str = helper(str);
            i ++;
        }

        return str;
    }
    
    public String helper(String str) {
        char tmp = str.charAt(0);
        int num = 1;
        StringBuffer sb = new StringBuffer("");
        
        for (int i = 1; i < str.length(); i ++) {
            if (str.charAt(i) == tmp) {
                num ++;
                continue;
            }
            
            sb.append(Integer.toString(num) + tmp);
            tmp = str.charAt(i);
            num = 1;
        }
        
        sb.append(Integer.toString(num) + tmp);
        
        return sb.toString();
    }
}

File: Python/038_CountAndSay.py
class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        result = "1"
        for i in range(n - 1):
            tmp = result
            result = ""
            cnt = 1
            for j in range(1, len(tmp)):
                if tmp[j] == tmp[j - 1]:
                    cnt += 1
                else:
                    result += ("%d"%cnt + tmp[j - 1])
                    cnt = 1
            result += ("%d"%cnt + tmp[len(tmp)-1])
            
        return result
                
            